File: Demo/Polymorphism/PY/main_class.py
# Polymorphism is a fancy term in programming that means "many forms."
# It allows you to treat objects of different classes in a similar way,
# even though they may have different underlying implementations.
# 1- Method Overriding
# There are a few reasons why Python doesn't support function overloading
# in the strict sense that languages like C++ do:
# Dynamic Typing: Python is a dynamically typed language.
# This means the data type of a variable is determined at runtime,
# not during compilation. This flexibility makes function overloading,
# which relies on static type checking, less straightforward in Python.


class Animal:
    def __init__(self, name):
        self.name = name

    def speak(self):
        print("Making animal sound...")


class Cat(Animal):
    def speak(self):
        print(f"{self.name} says Meow!")


class Dog(Animal):
    def speak(self):
        print(f"{self.name} says Woof!")


d = Dog("Fido")

c = Cat("Whiskers")

d.speak()

c.speak()

File: Demo/Polymorphism/CPP/mainClass.cpp
#include <iostream>

using namespace std;

// Method Overriding

class Animal
{
public:
    virtual void speak() = 0;
};

class Dog : public Animal
{
private:
    string name;

public:
    Dog(string n) : name(n)
    {
    }
    void speak() override
    {
        cout << name << " says Woof!" << endl;
    }
};

class Cat : public Animal
{
private:
    string name;

public:
    Cat(string n) : name(n)
    {
    }
    void speak() override
    {
        cout << name << " says Meow!" << endl;
    }
};

// Function Overloading:

int add(int num1, int num2);

double add(double num1, double num2);

int main()
{
    Animal *animal1 = new Dog("Fido");     // Animal pointer referencing a Dog object
    Animal *animal2 = new Cat("Whiskers"); // Animal pointer referencing a Cat object

    animal1->speak();
    animal2->speak();

    delete animal1;
    delete animal2;

    cout << add(2, 3) << endl;

    cout << add(2.5, 3.5) << endl;
}

int add(int num1, int num2)
{
    return num1 + num2;
}

double add(double num1, double num2)
{
    return num1 + num2;
}
